crates/autopilot/src/run_loop.rs

```diff
@@ -493,14 +495,66 @@ impl RunLoop {
     }
 }
 
+/// Checks whether or not an order's limit price is outside the market price
+/// specified by the quote.
+///
+/// Note that this check only looks at the order's limit price and the market
+/// price and is independent of amounts or trade direction.
+pub fn is_order_outside_market_price(
+    sell_amount: &U256,
+    buy_amount: &U256,
+    quote_buy_amount: U256,
+    quote_sell_amount: U256,
+) -> bool {
+    sell_amount.full_mul(quote_buy_amount) < quote_sell_amount.full_mul(*buy_amount)
+}
+
+/// Prepares the fee policies for each order in the auction.
+/// Determines if the protocol fee should be applied to the order.
+fn fee_policies(
+    auction: &Auction,
+    fee_policy: arguments::FeePolicy,
+    quotes: HashMap<OrderUid, Quote>,
+) -> HashMap<OrderUid, Vec<FeePolicy>> {
+    auction
+        .orders
+        .iter()
+        .filter_map(|order| {
+            match order.metadata.class {
+                OrderClass::Market => None,
+                OrderClass::Liquidity => None,
+                // TODO: https://github.com/cowprotocol/services/issues/2115
+                // skip protocol fee for TWAP limit orders
+                OrderClass::Limit(_) => {
+                    let quote = quotes.get(&order.metadata.uid)?;
+                    let quote_buy_amount = big_decimal_to_u256(&quote.buy_amount)?;
+                    let quote_sell_amount = big_decimal_to_u256(&quote.sell_amount)?;
```

Comment:
I see `fee_policies` function as part of the domain logic of autopilot. It should not be clogged with conversion of types. Let's move this conversion outside by making sure that `read_quotes` returns some `autopilot::Quote` struct with already prepared types for working with, rather than a `database::Quote`.

Reply:
I see that https://github.com/cowprotocol/services/pull/2157 is merged now. So will do the rework and move to the new file.